syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.metagameplay";
option java_outer_classname = "MetagameplayProto";
option objc_class_prefix = "MG";

package metagameplay;

service Metagameplayer {
  // Sends a greeting
  rpc SayHello (HelloRequest) returns (HelloReply) {}

  rpc SayHelloAgain (HelloRequest) returns (HelloReply) {}

  rpc Login (User) returns (SessionResponse) {}

  rpc Logout (User) returns (SessionResponse) {}

  rpc BuyItem (BuyRequest) returns (Status) {}

  rpc SellItem (BuyRequest) returns (Status) {}

  rpc GetAllItems (User) returns (ItemsResponse) {}

  rpc GetUser (User) returns (SessionResponse) {}
}

// The request message containing the user's name.
message HelloRequest {
  string name = 1;
}

// The response message containing the greetings
message HelloReply {
  string message = 1;
}

// items should be with out price
message User {
  string nickname = 1;
  string session = 2;
  uint32 credits = 3;
  repeated Item items = 4;
}

message Status {
  bool ok = 1;
  string error = 2;
}

message SessionResponse {
  User user = 1;
  Status status = 2;
}

message Item {
  string name = 1;
  uint32 price = 2;
}

message ItemsResponse {
  Status status = 1;
  repeated Item items = 2;
}

message BuyRequest {
  string item_name = 1;
  string session = 2;
}

